src/Microsoft.Management.Deployment/CatalogPackage.cpp

```diff
@@ -0,0 +1,99 @@
+#include "pch.h"
+#include <AppInstallerRepositorySource.h>
+#include <AppInstallerRepositorySearch.h>
+#include "CatalogPackage.h"
+#include "CatalogPackage.g.cpp"
+#include "PackageCatalog.h"
+#include "PackageVersionInfo.h"
+#include "PackageVersionId.h"
+#include <wil\cppwinrt_wrl.h>
+
+namespace winrt::Microsoft::Management::Deployment::implementation
+{
+    void CatalogPackage::Initialize(
+        std::shared_ptr<const ::AppInstaller::Repository::ISource> source,
+        std::shared_ptr<::AppInstaller::Repository::IPackage> package)
+    {
+        m_source = source;
+        m_package = package;
+    }
+    hstring CatalogPackage::Id()
+    {
+        return winrt::to_hstring(m_package->GetProperty(::AppInstaller::Repository::PackageProperty::Id).get());
+    }
+    hstring CatalogPackage::Name()
+    {
+        return winrt::to_hstring(m_package->GetProperty(::AppInstaller::Repository::PackageProperty::Name));
+    }
+    Microsoft::Management::Deployment::PackageVersionInfo CatalogPackage::InstalledVersion()
+    {
+        if (!m_installedVersion)
+        {
+            // InstalledVersion hasn't been created yet, create and populate it.
+            std::shared_ptr<::AppInstaller::Repository::IPackageVersion> installedVersion = m_package.get()->GetInstalledVersion();
+            if (installedVersion)
```

Comment:
>if (installedVersion) [](http://example.com/codeflow?start=12&length=21)

If there is no installed version we keep running this over and over.

Also, this isn't thread safe, but I'm not sure what model you gave this server.

Reply:
Applies generally to numerous instances below.  In addition to thread safety for the members of the class, we should ensure that our SQLite connection is in serialized mode.  We don't specify it currently, and while the default is serialized, I could not find documentation on which setting was used to compile winsqlite3.